samza-api/src/main/java/org/apache/samza/storage/kv/KeyValueIterable.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.storage.kv;
+
+public interface KeyValueIterable<K, V> extends Iterable<Entry<K, V>> {
```

Comment:
Does this need a "close" method so that you can close the internal iterator (e.g. tell internal RocksDB iterator to clean up snapshot)?

Reply:
Good question. I was thinking about this when I did the change. The reason that I didn't add the close method in KeyValueIterable is that the KeyValueIterator itself already has a close() method in the API, so it should be the responsibility of KeyValueIterator to close it (might even be confusing if we also have another close() here). Think of this as a factory class that creates the iterator, but close() should be on the iterator itself.

Reply:
I agree that KeyValueIterator should have the logic for closing the RocksDB snapshot. However, someone needs to call KeyValueIterator.close when it is done being used. If the KeyValueIterable uses the KeyValueIterator, then KeyValueIterable needs to provide a way for someone to call KeyValueIterator.close, or KeyValueIterable needs to call KeyValueIterator.close automatically.

Reply:
Oh, I believe when the users will get hold of KeyValueIterable instances when they call getIterator() from the iterable. So they can close the iterators by themselves.